django/db/models/indexes.py

```diff
@@ -0,0 +1,116 @@
+        errors = []
+        # Name shouldn't start with an underscore (Oracle hates this), so prepend D if we need to
+        if self._name[0] == '_':
+            errors.append('Index names cannot start with an underscore (_).')
+            self._name = 'D%s' % self._name[1:]
+        # It can't start with a number on Oracle, so prepend D if we need to
+        elif self._name[0].isdigit():
+            errors.append('Index names cannot start with a number (0-9).')
+            self._name = 'D%s' % self._name[1:]
+        return errors
+
+    def create_sql(self, schema_editor):
+        fields = [self.model._meta.get_field(field) for field in self.fields]
+        tablespace_sql = schema_editor._get_index_tablespace_sql(self.model, fields)
+        columns = [field.column for field in fields]
+
+        quote_name = schema_editor.quote_name
+        return schema_editor.sql_create_index % {
+            'table': quote_name(self.model._meta.db_table),
+            'name': quote_name(self.name),
+            'columns': ', '.join(quote_name(column) for column in columns),
+            'extra': tablespace_sql,
+        }
+
+    def remove_sql(self, schema_editor):
+        quote_name = schema_editor.quote_name
+        return schema_editor.sql_delete_index % {
+            'table': quote_name(self.model._meta.db_table),
+            'name': quote_name(self.name),
+        }
+
+    def deconstruct(self):
+        path = '%s.%s' % (self.__class__.__module__, self.__class__.__name__)
+        path = path.replace('django.db.models.indexes', 'django.db.models')
+        return (path, (), {'fields': self.fields})
+
+    @staticmethod
+    def _hash_generator(*args):
+        """
+        Generate a 32-bit digest of a set of arguments that can be used to
+        shorten identifying names.
+        """
+        h = hashlib.md5()
+        for arg in args:
+            h.update(force_bytes(arg))
+        return h.hexdigest()[:6]
+
+    def get_name(self, suffix='_idx'):
+        """
+        Generate a unique name for the index.
+
+        The name is divided into 3 parts - table name (12 chars), field name
+        (8 chars) and unique hash + suffix (10 chars). Each part is made to
+        fit its size by truncating the excess length.
+        """
+        table_name = self.model._meta.db_table
+        fields = [self.model._meta.get_field(field) for field in self.fields]
+        column_names = [field.column for field in fields]
+        hash_data = [table_name] + column_names + [self.index_type]
+        index_unique_hash = self._hash_generator(*hash_data)
+        table_part = table_name.replace('"', '').replace('.', '_')
+        field_part = column_names[0].replace('"', '').replace('.', '_')
+        hash_part = '%s%s' % (index_unique_hash, suffix)
+        if len(table_part) > 11:
+            table_part = table_part[:11]
+        if len(field_part) > 7:
+            field_part = field_part[:7]
+        if len(hash_part) > 10:
```

Comment:
In fact, suffixes aren't something input by users. They are something for us to decide. I wonder if there's a better way to keep future contributors informed about this restriction than raising an error.
While we are at this, do we really need customisation of suffixes. My idea is to keep different suffixes for indexes of different type like `_rt` for rtree, `_hsh` for hash, etc. which is exactly what I am planning to store in `index_type` in the future, so `get_name` doesn't need to have the `suffix` argument anymore.